File: src/sequence/tags.test.ts
import { describe, test, expect } from "vitest";
import type { SeqFactor } from "../types/Sequence";
import {
  filterByTag,
  extractTags,
  groupByTag,
  removeByTag,
  renameTag,
  excludeByTag,
  accumulateByTag,
} from "./tags";

const makeFactor = (
  id: string,
  tag: string,
  time: string,
  value: number,
  factor: "plus" | "minus" = "plus",
): SeqFactor => ({
  id,
  tag,
  time: new Date(time),
  value,
  factor,
});

const seq: SeqFactor[] = [
  makeFactor("1", "rent", "2024-01-01", 1000, "minus"),
  makeFactor("2", "food", "2024-01-05", 200, "minus"),
  makeFactor("3", "salary", "2024-01-10", 5000, "plus"),
  makeFactor("4", "food", "2024-01-15", 150, "minus"),
  makeFactor("5", "utilities", "2024-01-20", 100, "minus"),
  makeFactor("6", "rent", "2024-02-01", 1000, "minus"),
];

describe("filterByTag", () => {
  test("filters to matching tags", () => {
    const result = filterByTag(seq, ["food"]);
    expect(result).toHaveLength(2);
    expect(result.every((f) => f.tag === "food")).toBe(true);
  });

  test("filters to multiple tags", () => {
    const result = filterByTag(seq, ["rent", "utilities"]);
    expect(result).toHaveLength(3);
    expect(result.map((f) => f.id)).toEqual(["1", "5", "6"]);
  });

  test("returns empty array when no tags match", () => {
    expect(filterByTag(seq, ["nonexistent"])).toHaveLength(0);
  });

  test("returns empty array for empty sequence", () => {
    expect(filterByTag([], ["food"])).toHaveLength(0);
  });

  test("does not mutate the original array", () => {
    filterByTag(seq, ["food"]);
    expect(seq).toHaveLength(6);
  });
});

describe("extractTags", () => {
  test("returns unique tags", () => {
    const tags = extractTags(seq);
    expect(tags).toHaveLength(4);
    expect(new Set(tags)).toEqual(
      new Set(["rent", "food", "salary", "utilities"]),
    );
  });

  test("returns empty array for empty sequence", () => {
    expect(extractTags([])).toHaveLength(0);
  });

  test("returns single tag when all items share it", () => {
    const uniform = [
      makeFactor("1", "rent", "2024-01-01", 100),
      makeFactor("2", "rent", "2024-02-01", 100),
    ];
    expect(extractTags(uniform)).toEqual(["rent"]);
  });
});

describe("groupByTag", () => {
  test("groups factors by tag", () => {
    const groups = groupByTag(seq);
    expect(Object.keys(groups)).toHaveLength(4);
    expect(groups["rent"]).toHaveLength(2);
    expect(groups["food"]).toHaveLength(2);
    expect(groups["salary"]).toHaveLength(1);
    expect(groups["utilities"]).toHaveLength(1);
  });

  test("preserves factor data in groups", () => {
    const groups = groupByTag(seq);
    expect(groups["salary"][0].id).toBe("3");
    expect(groups["salary"][0].value).toBe(5000);
  });

  test("returns empty object for empty sequence", () => {
    expect(groupByTag([])).toEqual({});
  });
});

describe("excludeByTag", () => {
  test("excludes matching tags", () => {
    const result = excludeByTag(seq, ["food"]);
    expect(result).toHaveLength(4);
    expect(result.every((f) => f.tag !== "food")).toBe(true);
  });

  test("excludes multiple tags", () => {
    const result = excludeByTag(seq, ["food", "utilities"]);
    expect(result).toHaveLength(3);
    expect(result.map((f) => f.id)).toEqual(["1", "3", "6"]);
  });

  test("returns all when no tags match", () => {
    const result = excludeByTag(seq, ["nonexistent"]);
    expect(result).toHaveLength(6);
  });

  test("returns empty array for empty sequence", () => {
    expect(excludeByTag([], ["food"])).toHaveLength(0);
  });

  test("does not mutate the original array", () => {
    excludeByTag(seq, ["food"]);
    expect(seq).toHaveLength(6);
  });
});

describe("removeByTag", () => {
  test("is the same operation as excludeByTag", () => {
    const excluded = excludeByTag(seq, ["food", "rent"]);
    const removed = removeByTag(seq, ["food", "rent"]);
    expect(removed).toEqual(excluded);
  });
});

describe("renameTag", () => {
  test("renames all matching factors", () => {
    const result = renameTag(seq, "food", "dining");
    const diningItems = result.filter((f) => f.tag === "dining");
    const foodItems = result.filter((f) => f.tag === "food");
    expect(diningItems).toHaveLength(2);
    expect(foodItems).toHaveLength(0);
  });

  test("preserves other fields", () => {
    const result = renameTag(seq, "food", "dining");
    const renamed = result.find((f) => f.id === "2")!;
    expect(renamed.tag).toBe("dining");
    expect(renamed.value).toBe(200);
    expect(renamed.factor).toBe("minus");
  });

  test("leaves non-matching factors unchanged", () => {
    const result = renameTag(seq, "food", "dining");
    expect(result.find((f) => f.id === "1")!.tag).toBe("rent");
  });

  test("returns unchanged copy when tag not found", () => {
    const result = renameTag(seq, "nonexistent", "new");
    expect(result).toHaveLength(6);
    expect(result).not.toBe(seq);
    expect(result).toEqual(seq);
  });

  test("does not mutate the original array", () => {
    renameTag(seq, "food", "dining");
    expect(seq[1].tag).toBe("food");
  });
});

describe("accumulateByTag", () => {
  test("accumulates only the specified tag", () => {
    const result = accumulateByTag(seq, 0, "food");
    expect(result).toHaveLength(2);
    expect(result[0].store).toBe(-200);
    expect(result[1].store).toBe(-350);
  });

  test("returns empty array when tag not found", () => {
    expect(accumulateByTag(seq, 0, "nonexistent")).toEqual([]);
  });

  test("applies base value", () => {
    const result = accumulateByTag(seq, 1000, "rent");
    expect(result[0].store).toBe(0); // 1000 - 1000
    expect(result[1].store).toBe(-1000); // 0 - 1000
  });

  test("returns empty array for empty sequence", () => {
    expect(accumulateByTag([], 0, "food")).toEqual([]);
  });
});

File: tags_test.go
package aggrett

import (
	"reflect"
	"testing"
)

func makeTagFactor(t *testing.T, id, tag, timeValue string, value float64, factor Factor) SeqFactor {
	t.Helper()
	if factor == "" {
		factor = FactorPlus
	}
	return SeqFactor{ID: id, Tag: tag, Time: mustParseDate(t, timeValue), Value: value, Factor: factor}
}

func TestFilterByTag(t *testing.T) {
	seq := []SeqFactor{
		makeTagFactor(t, "1", "rent", "2024-01-01", 1000, FactorMinus),
		makeTagFactor(t, "2", "food", "2024-01-05", 200, FactorMinus),
		makeTagFactor(t, "3", "salary", "2024-01-10", 5000, FactorPlus),
		makeTagFactor(t, "4", "food", "2024-01-15", 150, FactorMinus),
		makeTagFactor(t, "5", "utilities", "2024-01-20", 100, FactorMinus),
		makeTagFactor(t, "6", "rent", "2024-02-01", 1000, FactorMinus),
	}

	t.Run("filters to matching tags", func(t *testing.T) {
		result := FilterByTag(seq, []string{"food"})
		if len(result) != 2 {
			t.Fatalf("got %d items want 2", len(result))
		}
		for _, f := range result {
			if f.Tag != "food" {
				t.Fatalf("unexpected tag %q", f.Tag)
			}
		}
	})

	t.Run("filters to multiple tags", func(t *testing.T) {
		result := FilterByTag(seq, []string{"rent", "utilities"})
		if len(result) != 3 || result[0].ID != "1" || result[1].ID != "5" || result[2].ID != "6" {
			t.Fatalf("unexpected result %#v", result)
		}
	})

	t.Run("returns empty array when no tags match", func(t *testing.T) {
		if len(FilterByTag(seq, []string{"nonexistent"})) != 0 {
			t.Fatalf("expected empty result")
		}
	})

	t.Run("returns empty array for empty sequence", func(t *testing.T) {
		if len(FilterByTag(nil, []string{"food"})) != 0 {
			t.Fatalf("expected empty result")
		}
	})

	t.Run("does not mutate the original array", func(t *testing.T) {
		_ = FilterByTag(seq, []string{"food"})
		if len(seq) != 6 {
			t.Fatalf("input sequence mutated")
		}
	})
}

func TestExtractTags(t *testing.T) {
	seq := []SeqFactor{
		makeTagFactor(t, "1", "rent", "2024-01-01", 1000, FactorMinus),
		makeTagFactor(t, "2", "food", "2024-01-05", 200, FactorMinus),
		makeTagFactor(t, "3", "salary", "2024-01-10", 5000, FactorPlus),
		makeTagFactor(t, "4", "food", "2024-01-15", 150, FactorMinus),
		makeTagFactor(t, "5", "utilities", "2024-01-20", 100, FactorMinus),
		makeTagFactor(t, "6", "rent", "2024-02-01", 1000, FactorMinus),
	}

	t.Run("returns unique tags", func(t *testing.T) {
		tags := ExtractTags(seq)
		if len(tags) != 4 {
			t.Fatalf("got %d tags want 4", len(tags))
		}
		if !reflect.DeepEqual(tags, []string{"rent", "food", "salary", "utilities"}) {
			t.Fatalf("unexpected tags %#v", tags)
		}
	})

	t.Run("returns empty array for empty sequence", func(t *testing.T) {
		if len(ExtractTags(nil)) != 0 {
			t.Fatalf("expected empty tags")
		}
	})

	t.Run("returns single tag when all items share it", func(t *testing.T) {
		uniform := []SeqFactor{
			makeTagFactor(t, "1", "rent", "2024-01-01", 100, ""),
			makeTagFactor(t, "2", "rent", "2024-02-01", 100, ""),
		}
		tags := ExtractTags(uniform)
		if !reflect.DeepEqual(tags, []string{"rent"}) {
			t.Fatalf("unexpected tags %#v", tags)
		}
	})
}

func TestGroupByTag(t *testing.T) {
	seq := []SeqFactor{
		makeTagFactor(t, "1", "rent", "2024-01-01", 1000, FactorMinus),
		makeTagFactor(t, "2", "food", "2024-01-05", 200, FactorMinus),
		makeTagFactor(t, "3", "salary", "2024-01-10", 5000, FactorPlus),
		makeTagFactor(t, "4", "food", "2024-01-15", 150, FactorMinus),
		makeTagFactor(t, "5", "utilities", "2024-01-20", 100, FactorMinus),
		makeTagFactor(t, "6", "rent", "2024-02-01", 1000, FactorMinus),
	}

	t.Run("groups factors by tag", func(t *testing.T) {
		groups := GroupByTag(seq)
		if len(groups) != 4 {
			t.Fatalf("got %d groups want 4", len(groups))
		}
		if len(groups["rent"]) != 2 || len(groups["food"]) != 2 || len(groups["salary"]) != 1 || len(groups["utilities"]) != 1 {
			t.Fatalf("unexpected groups %#v", groups)
		}
	})

	t.Run("preserves factor data in groups", func(t *testing.T) {
		groups := GroupByTag(seq)
		if groups["salary"][0].ID != "3" || groups["salary"][0].Value != 5000 {
			t.Fatalf("unexpected salary group %#v", groups["salary"])
		}
	})

	t.Run("returns empty object for empty sequence", func(t *testing.T) {
		groups := GroupByTag(nil)
		if len(groups) != 0 {
			t.Fatalf("expected empty groups")
		}
	})
}

func TestExcludeByTag(t *testing.T) {
	seq := []SeqFactor{
		makeTagFactor(t, "1", "rent", "2024-01-01", 1000, FactorMinus),
		makeTagFactor(t, "2", "food", "2024-01-05", 200, FactorMinus),
		makeTagFactor(t, "3", "salary", "2024-01-10", 5000, FactorPlus),
		makeTagFactor(t, "4", "food", "2024-01-15", 150, FactorMinus),
		makeTagFactor(t, "5", "utilities", "2024-01-20", 100, FactorMinus),
		makeTagFactor(t, "6", "rent", "2024-02-01", 1000, FactorMinus),
	}

	t.Run("excludes matching tags", func(t *testing.T) {
		result := ExcludeByTag(seq, []string{"food"})
		if len(result) != 4 {
			t.Fatalf("got %d items want 4", len(result))
		}
		for _, f := range result {
			if f.Tag == "food" {
				t.Fatalf("food tag should have been excluded")
			}
		}
	})

	t.Run("excludes multiple tags", func(t *testing.T) {
		result := ExcludeByTag(seq, []string{"food", "utilities"})
		if len(result) != 3 || result[0].ID != "1" || result[1].ID != "3" || result[2].ID != "6" {
			t.Fatalf("unexpected result %#v", result)
		}
	})

	t.Run("returns all when no tags match", func(t *testing.T) {
		result := ExcludeByTag(seq, []string{"nonexistent"})
		if len(result) != 6 {
			t.Fatalf("got %d items want 6", len(result))
		}
	})

	t.Run("returns empty array for empty sequence", func(t *testing.T) {
		if len(ExcludeByTag(nil, []string{"food"})) != 0 {
			t.Fatalf("expected empty result")
		}
	})

	t.Run("does not mutate the original array", func(t *testing.T) {
		_ = ExcludeByTag(seq, []string{"food"})
		if len(seq) != 6 {
			t.Fatalf("input sequence mutated")
		}
	})
}

func TestRemoveByTag(t *testing.T) {
	seq := []SeqFactor{
		makeTagFactor(t, "1", "rent", "2024-01-01", 1000, FactorMinus),
		makeTagFactor(t, "2", "food", "2024-01-05", 200, FactorMinus),
		makeTagFactor(t, "3", "salary", "2024-01-10", 5000, FactorPlus),
	}

	excluded := ExcludeByTag(seq, []string{"food", "rent"})
	removed := RemoveByTag(seq, []string{"food", "rent"})
	if !reflect.DeepEqual(removed, excluded) {
		t.Fatalf("removeByTag mismatch: removed=%#v excluded=%#v", removed, excluded)
	}
}

func TestRenameTag(t *testing.T) {
	seq := []SeqFactor{
		makeTagFactor(t, "1", "rent", "2024-01-01", 1000, FactorMinus),
		makeTagFactor(t, "2", "food", "2024-01-05", 200, FactorMinus),
		makeTagFactor(t, "3", "salary", "2024-01-10", 5000, FactorPlus),
		makeTagFactor(t, "4", "food", "2024-01-15", 150, FactorMinus),
		makeTagFactor(t, "5", "utilities", "2024-01-20", 100, FactorMinus),
		makeTagFactor(t, "6", "rent", "2024-02-01", 1000, FactorMinus),
	}

	t.Run("renames all matching factors", func(t *testing.T) {
		result := RenameTag(seq, "food", "dining")
		diningCount := 0
		foodCount := 0
		for _, f := range result {
			if f.Tag == "dining" {
				diningCount++
			}
			if f.Tag == "food" {
				foodCount++
			}
		}
		if diningCount != 2 || foodCount != 0 {
			t.Fatalf("unexpected tag counts dining=%d food=%d", diningCount, foodCount)
		}
	})

	t.Run("preserves other fields", func(t *testing.T) {
		result := RenameTag(seq, "food", "dining")
		var renamed SeqFactor
		for _, f := range result {
			if f.ID == "2" {
				renamed = f
				break
			}
		}
		if renamed.Tag != "dining" || renamed.Value != 200 || renamed.Factor != FactorMinus {
			t.Fatalf("unexpected renamed factor %#v", renamed)
		}
	})

	t.Run("leaves non-matching factors unchanged", func(t *testing.T) {
		result := RenameTag(seq, "food", "dining")
		for _, f := range result {
			if f.ID == "1" && f.Tag != "rent" {
				t.Fatalf("non-matching factor changed %#v", f)
			}
		}
	})

	t.Run("returns unchanged copy when tag not found", func(t *testing.T) {
		result := RenameTag(seq, "nonexistent", "new")
		if len(result) != 6 {
			t.Fatalf("got %d items want 6", len(result))
		}
		if !reflect.DeepEqual(result, seq) {
			t.Fatalf("expected unchanged copy")
		}
		if &result[0] == &seq[0] {
			t.Fatalf("expected copy, got same backing elements")
		}
	})

	t.Run("does not mutate the original array", func(t *testing.T) {
		_ = RenameTag(seq, "food", "dining")
		if seq[1].Tag != "food" {
			t.Fatalf("input sequence mutated")
		}
	})
}

func TestAccumulateByTag(t *testing.T) {
	seq := []SeqFactor{
		makeTagFactor(t, "1", "rent", "2024-01-01", 1000, FactorMinus),
		makeTagFactor(t, "2", "food", "2024-01-05", 200, FactorMinus),
		makeTagFactor(t, "3", "salary", "2024-01-10", 5000, FactorPlus),
		makeTagFactor(t, "4", "food", "2024-01-15", 150, FactorMinus),
		makeTagFactor(t, "5", "utilities", "2024-01-20", 100, FactorMinus),
		makeTagFactor(t, "6", "rent", "2024-02-01", 1000, FactorMinus),
	}

	t.Run("accumulates only the specified tag", func(t *testing.T) {
		result := AccumulateByTag(seq, 0, "food")
		if len(result) != 2 || result[0].Store != -200 || result[1].Store != -350 {
			t.Fatalf("unexpected result %#v", result)
		}
	})

	t.Run("returns empty array when tag not found", func(t *testing.T) {
		result := AccumulateByTag(seq, 0, "nonexistent")
		if len(result) != 0 {
			t.Fatalf("expected empty result")
		}
	})

	t.Run("applies base value", func(t *testing.T) {
		result := AccumulateByTag(seq, 1000, "rent")
		if len(result) != 2 || result[0].Store != 0 || result[1].Store != -1000 {
			t.Fatalf("unexpected result %#v", result)
		}
	})

	t.Run("returns empty array for empty sequence", func(t *testing.T) {
		result := AccumulateByTag(nil, 0, "food")
		if len(result) != 0 {
			t.Fatalf("expected empty result")
		}
	})
}
